### PR TITLE
Remove legacy device arguments in Trainer

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Removes
```py
Trainer(gpus=...)
Trainer(tpu_cores=...)
Trainer(ipus=...)
```
The new way since 1.5 is
```
Trainer(devices=...)
```

### Does your PR introduce any breaking changes? If yes, please list them.

Yes

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -225,20 +221,14 @@ def __init__(
                 that don't support deterministic mode (requires PyTorch 1.11+). If not set, defaults to ``False``.
                 Default: ``None``.
 
-            devices: Will be mapped to either `gpus`, `tpu_cores`, `num_processes` or `ipus`,
-                based on the accelerator type.
+            devices: The devices to use. Can be set to a positive number (int or str), a sequence of device indices
+                (list or str), the value ``-1`` to indicate all available devices should be used, or ``"auto"`` for
+                automatic selection based on the chosen accelerator. Default: ``"auto"``.
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/trainer.py:226]: Isn't the default 1?

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/models/test_tpu.py [hunk 2]
```diff
@@ -259,12 +259,15 @@ def test_accelerator_set_when_using_tpu(devices):
 
 @pytest.mark.parametrize(
     ["cli_args", "expected"],
-    [("--tpu_cores=8", {"tpu_cores": 8}), ("--tpu_cores=1,", {"tpu_cores": "1,"})],
+    [
+        ("--accelerator=tpu --devices=8", {"accelerator": "tpu", "devices": 8}),
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_tpu.py:263]: Standalone tests seem to have an issue with the whitespace here between the flags.

You can try to fix the space escaping in the script
or use `pytest.param(..., id="foo")`

**Review by justusschock — APPROVED** [review]

File: tests/tests_pytorch/models/test_tpu.py (hunk 2 above)
Comment (awaelchli) [reply tests/tests_pytorch/models/test_tpu.py]: Interesting, how did you find this? All I saw in the logs was "stopping because of potential error".

File: src/pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:230]: No, it is auto:

```py
    def _set_devices_flag_if_auto_passed(self) -> None:
        if self._devices_flag == "auto" or self._devices_flag is None:
            self._devices_flag = self.accelerator.auto_device_count()
```

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:230]: We can consider making a follow up that sets this explicitly in the constructor as the default value rather than None

Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:230]: It says:

```py
ERROR: not found: /__w/2/s/tests/tests_pytorch/models/test_tpu.py::test_tpu_devices_with_argparse[--accelerator=tpu
(no name '/__w/2/s/tests/tests_pytorch/models/test_tpu.py::test_tpu_devices_with_argparse[--accelerator=tpu' in any of [<DoctestModule tests/tests_pytorch/models/test_tpu.py>, <Module tests/tests_pytorch/models/test_tpu.py>])
```

File: tests/tests_pytorch/models/test_tpu.py (hunk 2 above)
Comment (carmocca) [reply tests/tests_pytorch/models/test_tpu.py]: It says:

```py
ERROR: not found: /__w/2/s/tests/tests_pytorch/models/test_tpu.py::test_tpu_devices_with_argparse[--accelerator=tpu
(no name '/__w/2/s/tests/tests_pytorch/models/test_tpu.py::test_tpu_devices_with_argparse[--accelerator=tpu' in any of [<DoctestModule tests/tests_pytorch/models/test_tpu.py>, <Module tests/tests_pytorch/models/test_tpu.py>])
```

File: src/pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:230]: That would be good.
